torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2357,103 +2357,323 @@ Tensor slice_backward_wrapper(
   return slice_backward(grad, input_sizes, dim, start_val, end_val, step);
 }
 
-// https://j-towns.github.io/papers/svd-derivative.pdf
-//
-// This makes no assumption on the signs of sigma.
-Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads,
-          bool full_matrices, const Tensor& U, const Tensor& sigma, const Tensor& Vh) {
-  auto m = U.size(-2);
-  auto n = Vh.size(-1);
-  auto k = sigma.size(-1);
-  auto gsigma = grads[1];
-
-  auto u = U;
-  auto v = Vh.mH();
-  auto gu = grads[0];
-  // grads[2] has the gradient wrt Vh
-  auto gv = grads[2].defined() ? grads[2].mH() : grads[2];
-
-  if (full_matrices) {
-    // We ignore the free subspace here because possible base vectors cancel
-    // each other, e.g., both -v and +v are valid base for a dimension.
-    // Don't assume behavior of any particular implementation of svd.
-    u = u.narrow(-1, 0, k);
-    v = v.narrow(-1, 0, k);
-    if (gu.defined()) {
-      gu = gu.narrow(-1, 0, k);
+std::tuple<Tensor, Tensor, Tensor> linalg_svd_jvp(const Tensor& dA,
+                                                  const Tensor& U_,
+                                                  const Tensor& S,
+                                                  const Tensor& Vh_,
+                                                  const bool full_matrices) {
+  // See svd_backward for the derivation
+  // With sym(X) = X + X^H, we implement
+  // dU = U (sym(dX S) / E + i Im(diag(dX)) / (2S))
+  // if m > n
+  //   dU = [dU for m == n] + (I_m - UU^H) dA V S^{-1}
+  // dS = Re(diag(dP))
+  // dV = V (sym(S dX) / E - i Im(diag(dX)) / (2S))
+  // if m < n
+  //   dV = [dV for m == n] + (I_n - VV^H) (dA)^H U S^{-1}
+  // dVh = dV^H
+  // with dP = U^H dA V
+  //      dX = dP - dS
+  //      E_{jk} = (S_j - S_k)
+  //      E_{kk} = 1
+  const auto is_complex = dA.is_complex();
+  const auto m = dA.size(-2);
+  const auto n = dA.size(-1);
+  const auto k = S.size(-1);
+
+  const auto U = full_matrices ? U_.narrow(-1, 0, k) : U_;
+  const auto Vh = full_matrices ? Vh_.narrow(-2, 0, k) : Vh_;
+  const auto V = Vh.mH();
+
+  // dP = U^H dA V
+  const auto dP = m >= n ? at::matmul(U.mH(), at::matmul(dA, V))
+                         : at::matmul(at::matmul(U.mH(), dA), V);
+
+  // We don't want dS to be a view into a larger tensor so we clone it
+  auto dS = is_complex ? at::real(dP.diagonal(0, -2, -1)).clone()
+                       : dP.diagonal(0, -2, -1).clone();
+
+  // dX = dP - dS
+  // Here we update dP in-place rather than
+  if (is_complex) {
+    at::real(dP.diagonal(0, -2, -1)).zero_();
+  } else {
+    dP.diagonal(0, -2, -1).zero_();
+  }
+
+  auto E = [&S]{
+    const auto S2 = S * S;
+    auto ret = S2.unsqueeze(-2) - S2.unsqueeze(-1);
+    // Any number a != 0 would, as we are going to compute 0 / a
+    ret.diagonal(0, -2, -1).fill_(1);
+    return ret;
+  }();
+
+  const auto sym = [](const Tensor& X) { return X + X.mH(); };
```

Comment:
Usually symmetrization is done with `X -> (X + X.mH()) / 2` so that if X is already a Hermitian matrix it's not changed. Is omitting 1/2 factor done on purpose here?
Tests pass so everything is good then.